packages/decoder/lib/ProviderAdapter.ts

```diff
@@ -0,0 +1,167 @@
+import type { BlockSpecifier, RegularizedBlockSpecifier } from "./types";
+import type BN from "bn.js";
+import type { Provider } from "web3/providers";
+import { promisify } from "util";
+
+// lifted from @types/web3
+type Log = {
+    address: string;
+    data: string;
+    topics: string[];
+    logIndex: number;
+    transactionHash: string;
+    transactionIndex: number;
+    blockHash: string;
+    blockNumber: number;
+}
+type PastLogsOptions = {
+  toBlock?: string | number;
+  fromBlock?: string | number;
+  address?: string | string[];
+};
+type SendRequestArgs = {
+  method: string;
+  params: unknown[];
+};
+type EipElevenNinetyThreeProvider = {
+  request: (options: { method: string; params: unknown[]; }) => Promise<any>;
+}
+const stringWhitelist = [
+  "latest",
+  "pending",
+  "genesis",
+  "earliest"
+];
+
+const formatBlockSpecifier = (block: BlockSpecifier): string => {
+  if (typeof block === "string" && stringWhitelist.includes(block)) {
+    // block is one of 'latest', 'pending', 'earliest', or 'genesis'
+    return block === "genesis" ?
+      // convert old web3 input format which uses 'genesis'
+      "earliest" :
+      block;
+  } else if (typeof block === "string" && !isNaN(parseInt(block))) {
+    // block is a string representation of a number
+    if (block.startsWith("0x")) return block;
+    // convert to hex and add '0x' prefix in case block is decimal
+    return `0x${parseInt(block).toString(16)}`;
+  } else if (typeof block === "number") {
+    return `0x${block.toString(16)}`;
+  } else {
+    throw new Error(
+      "The block specified must be a number or one of the strings 'latest'," +
```

Comment:
Here I decided to neglect mentioning "genesis", should it be mentioned here? I figured it would encourage people to use the rpc spec terms.